Source/JavaScriptCore/builtins/MapConstructor.js

```diff
@@ -27,12 +27,16 @@ function groupBy(items, callback)
 {
     "use strict";
 
-    if (!@isObject(items))
-        @throwTypeError("Map.groupBy requires that the first argument must be an object");
+    if (@isUndefinedOrNull(items))
+        @throwTypeError("Map.groupBy requires that the first argument not to be null or undefined");
 
     if (!@isCallable(callback))
         @throwTypeError("Map.groupBy requires that the second argument must be a function");
 
+    var iteratorMethod = items.@@iterator;
+    if (!@isCallable(iteratorMethod))
+        @throwTypeError("Map.groupBy requires that the property of the first argument, items[Symbol.iterator] be a function");
```

Comment:
Even without this change, TypeError will still be thrown, but the error message will be `TypeError: undefined is not a function`. This is not user friendly. I would like to know if there is a better way.

Reply:
I think this is semantically wrong as if `@@iterator` is a  getter then it will be run twice. You could do:
```js
    var iteratorMethod = items.@@iterator;
    if (!@isCallable(iteratorMethod))
        @throwTypeError("Map.groupBy requires that the property of the first argument, items[Symbol.iterator] be a function");
    var iterator = iteratorMethod.@call(items);

    // Since for-of loop once more looks up the @@iterator property of a given iterable,
    // it could be observable if the user defines a getter for @@iterator.
    // To avoid this situation, we define a wrapper object that @@iterator just returns a given iterator.
    var wrapper = {
        @@iterator: function () { return iterator; }
    };

    for (var item of wrapper) {
```

Reply:
Thanks for the review! That snippet is very interesting. I have fixed all the you reviewed and will ask you to review it again.